clients/java/src/main/java/io/camunda/zeebe/client/impl/worker/JobStreamerImpl.java

```diff
@@ -0,0 +1,201 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.zeebe.client.impl.worker;
+
+import io.camunda.zeebe.client.api.ZeebeFuture;
+import io.camunda.zeebe.client.api.command.FinalCommandStep;
+import io.camunda.zeebe.client.api.command.StreamJobsCommandStep1.StreamJobsCommandStep3;
+import io.camunda.zeebe.client.api.response.ActivatedJob;
+import io.camunda.zeebe.client.api.response.StreamJobsResponse;
+import io.camunda.zeebe.client.api.worker.BackoffSupplier;
+import io.camunda.zeebe.client.api.worker.JobClient;
+import io.camunda.zeebe.client.impl.Loggers;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+import net.jcip.annotations.GuardedBy;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+
+@ThreadSafe
+final class JobStreamerImpl implements JobStreamer {
+  private static final Logger LOGGER = Loggers.JOB_WORKER_LOGGER;
+
+  private final JobClient jobClient;
+  private final String jobType;
+  private final String workerName;
+  private final Duration timeout;
+  private final List<String> fetchVariables;
+  private final Duration requestTimeout;
+  private final BackoffSupplier backoffSupplier;
+  private final ScheduledExecutorService executor;
+  private final Lock streamLock;
+
+  @GuardedBy("streamLock")
+  private ZeebeFuture<StreamJobsResponse> streamControl;
+
+  @GuardedBy("streamLock")
+  private FinalCommandStep<StreamJobsResponse> command;
+
+  @GuardedBy("streamLock")
+  private boolean isClosed;
+
+  @GuardedBy("streamLock")
+  private long retryDelay;
+
+  public JobStreamerImpl(
+      final JobClient jobClient,
+      final String jobType,
+      final String workerName,
+      final Duration timeout,
+      final List<String> fetchVariables,
+      final Duration requestTimeout,
+      final BackoffSupplier backoffSupplier,
+      final ScheduledExecutorService executor) {
+    this.jobClient = jobClient;
+    this.jobType = jobType;
+    this.workerName = workerName;
+    this.timeout = timeout;
+    this.fetchVariables = fetchVariables;
+    this.requestTimeout = requestTimeout;
+    this.backoffSupplier = backoffSupplier;
+    this.executor = executor;
+
+    streamLock = new ReentrantLock();
+  }
+
+  @Override
+  public void close() {
+    try {
+      streamLock.lockInterruptibly();
```

Comment:
❓ How do we benefit allowing thread interruption?

Reply:
When writing any code which may get integrated into someone else's project (e.g. library code), it's best to be a "good" citizen. If we did lock interruptibly, it means that we would try to grab this lock "forever". So if someone is using our client in their application, and they want to do some graceful shutdown which includes closing the streamer, they wouldn't be able to without forcefully killing the thread.

So instead, a good rule of thumb is to always account for possible interrupt exception, as it's the mechanism applications use to tell a thread it should finish its work and stop.

Here is a good example; we don't want to block forever. Another example is when you have a classic run loop for a daemon thread or the like, where you would do:

```java
while (true) {
  doWork();
} 
```

There's no way for anyone outside to terminate this. Instead, you do:

```java
final var thread = Thread.current();
while (!thread.isInterrupted()) {
  doSomeWork();
}
```

Reply:
Very informative explanation, totally makes sense! 🚀